File: nimbusvault/shared/auth_middleware.js
const fs = require('fs');
const path = require('path');
const jwt = require('jsonwebtoken');

const publicKey = fs.readFileSync(path.join(__dirname, '../auth-service/keys/public.pem'), 'utf8');

function authMiddleware(req, res, next) {
  let token = null;
  if (req.headers.authorization && req.headers.authorization.startsWith('Bearer ')) {
    token = req.headers.authorization.split(' ')[1];
  }
  if (!token && req.cookies) {
    token = req.cookies['access_token'];
  }
  if (!token) {
    return res.status(401).json({ message: 'No token provided' });
  }
  try {
    const payload = jwt.verify(token, publicKey, { algorithms: ['RS256'] });
    req.user = payload;
    return next();
  } catch (err) {
    return res.status(401).json({ message: 'Invalid token' });
  }
}

module.exports = authMiddleware;

File: nimbusvault/shared/auth_middleware.py
from pathlib import Path
from fastapi import Request, HTTPException
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
import jwt

class AuthMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, public_key: str):
        super().__init__(app)
        self.public_key = public_key

    async def dispatch(self, request: Request, call_next):
        token = None
        auth = request.headers.get('Authorization')
        if auth and auth.lower().startswith('bearer '):
            token = auth.split(' ', 1)[1]
        if not token:
            token = request.cookies.get('access_token')
        if token:
            try:
                payload = jwt.decode(token, self.public_key, algorithms=['RS256'])
                request.state.user = payload
            except jwt.PyJWTError:
                return JSONResponse(status_code=401, content={'detail': 'Invalid token'})
        else:
            request.state.user = None
        return await call_next(request)

def get_current_user(request: Request):
    if not getattr(request.state, 'user', None):
        raise HTTPException(status_code=401, detail='Not authenticated')
    return request.state.user

